From 631893f750746fe7d873ab6dcd63a41ac66fad0a Mon Sep 17 00:00:00 2001
From: Christoph Reiter <reiter.christoph@gmail.com>
Date: Sun, 5 Aug 2018 11:21:26 +0200
Subject: [PATCH] gstdio: use _stat64 for GStatBuf on 64bit mingw. Fixes #1476

The size of stat depends on various macros on Windows which leads to
the problem of size mismatches when glib is built with a different configuration
than a program using it.

For example the autotools build defaults to _FILE_OFFSET_BITS=64 and a program
not defining  _FILE_OFFSET_BITS will allocate a too small struct on the stack,
leading to stack corruption when glib writes to it.

To make the size the user sees always match the default mingw build define GStatBuf
as _stat64 (same as _FILE_OFFSET_BITS=64) under mingw+64bit.
---
 glib/gstdio.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/glib/gstdio.h b/glib/gstdio.h
index 15091b3e7..94f4fa36f 100644
--- a/glib/gstdio.h
+++ b/glib/gstdio.h
@@ -44,6 +44,10 @@ G_BEGIN_DECLS
 
 typedef struct _stat32 GStatBuf;
 
+#elif defined(__MINGW64_VERSION_MAJOR) && defined(_WIN64)
+
+typedef struct _stat64 GStatBuf;
+
 #else
 
 typedef struct stat GStatBuf;
-- 
2.17.1

